correctly skip docs incorrectly being read in to slices

gabrielrussell opened the pull request:
When trying to parse a bson document as a slice, and the slice is not of DocElem or RawDocElem, readElemTo would decide to do nothing, but didn't consume the element, leaving the decoder in a bad state.

niemeyer commented:
Hey Gabriel,

This doesn't look quite right, and running tests confirms it is indeed introducing a bug. The new logic should be in default, shouldn't it?  Have you tried to run tests there?

Also, we need a test demonstrating why the current logic is broken, and to prevent regressions.

Thanks for the fix.

Commit:
correctly skip docs being read in to slices

o when unmarshaling a document in to a slice the parser forgot
  to discard the document, and left the parser in a broken state.
  Now it's properly discarded
o test that unmarshaling a document in to a slice doesn't error

gabrielrussell commented:
Yes, absolutely, in the default. I've also added a test.

niemeyer commented:
Thanks Gabriel.
